visual-js/visual-playwright/src/api.ts

```diff
@@ -0,0 +1,325 @@
+  const { buildId } = getOpts();
+
+  if (!buildId) {
+    return;
+  }
+
+  const browser = page.context().browser();
+  const {
+    screenshotOptions = {},
+    clipSelector,
+    delay,
+    captureDom,
+    ignoreRegions: userIgnoreRegions,
+    diffingMethod,
+  } = options ?? {};
+  const { animations = 'disabled', caret } = screenshotOptions;
+  let ignoreRegions: RegionIn[] = [];
+
+  const promises: Promise<unknown>[] = [
+    // Wait for all fonts to be loaded & ready
+    page.evaluate(() => document.fonts.ready),
+  ];
+
+  if (delay) {
+    // If a delay has been configured by the user, append it to our promises
+    promises.push(new Promise((resolve) => setTimeout(resolve, delay)));
+  }
+
+  if (userIgnoreRegions) {
+    promises.push(
+      (async (): Promise<void> => {
+        const filterIgnoreRegion = (
+          region: RegionIn | string,
+        ): region is RegionIn => typeof region !== 'string';
+        const filterIgnoreSelector = (
+          region: RegionIn | string,
+        ): region is string => typeof region === 'string';
+
+        const selectors = userIgnoreRegions.filter(filterIgnoreSelector);
+        let selectorRegions: RegionIn[] = [];
+
+        if (selectors.length) {
+          selectorRegions = await page.evaluate(
+            ({ selectors }) => {
+              const selectorRegions: RegionIn[] = [];
+              selectors.forEach((selector) => {
+                const elements = document.querySelectorAll(selector);
+                elements.forEach((element) => {
+                  const rect = element.getBoundingClientRect();
+
+                  selectorRegions.push({
+                    name: selector,
+                    x: Math.round(rect.x),
+                    y: Math.round(rect.y),
+                    height: Math.round(rect.height),
+                    width: Math.round(rect.width),
+                  });
+                });
+              });
+              return selectorRegions;
+            },
+            { selectors },
+          );
+        }
+
+        ignoreRegions = [
+          ...userIgnoreRegions.filter(filterIgnoreRegion),
+          ...selectorRegions,
+        ].filter((region) => 0 < region.width * region.height);
+      })(),
+    );
+  }
+
+  // Await all queued / concurrent promises before resuming
+  await Promise.all(promises);
+
+  const clip = clipSelector
+    ? await page.evaluate(
+        ({ clipSelector }) => {
+          // Clip the screenshot to the dims of the requested clipSelector.
+          const clipElement = document.querySelector(clipSelector);
+          if (!clipElement) {
+            return undefined;
+          }
+
+          const clientDims = clipElement.getBoundingClientRect();
+          const { x, y, height, width } = clientDims;
+
+          // If any values are falsy (or zero), then do not clip as those are invalid options for
+          // playwright
+          if (x < 0 || y < 0 || !height || !width) {
```

Comment:
Here shouldn't we use `Math.max(x, 0)` when `x+height > 0` ? Same applies to `y` coordinates.

If clipSelector is partially out of viewport (with x-coord negative), shouldn't we crop the visible area ?

Reply:
Yeah, I think that would work.